File: 2023/day01.cpp
#include <iostream>
#include <fstream>

using namespace std;

int getNum(string line) {
    int num = 0;

    for(char c : line) {
        if(isdigit(c)) {
            num += c - '0';
            break;
        }
    }

    num *= 10;

    for(int i = line.size() - 1; i >= 0; i--) {
        if(isdigit(line[i])) {
            num += line[i] - '0';
            break;
        }
    }

    return num;
}

int main() {
    ifstream file("day01.txt");
    string line;
    int ans = 0;
    while(getline(file, line)) {
        ans += getNum(line);
    }
    cout << ans << "\n";
}

File: 2023/day01.rs
use std::fs;
use std::num::ParseIntError;

fn get_num(line: String) -> Result<i32, ParseIntError> {
    let num: String = line.chars().filter(|c| c.is_numeric()).collect();

    let first = num.chars().next().unwrap();
    let last = num.chars().last().unwrap();
    format!("{}{}", first, last).parse()
}

fn main() {
    let file = fs::read_to_string("day01.txt").expect("No such file");

    let mut sum: i32 = 0;
    for line in file.lines() {
        let num = match get_num(line.to_string()) {
            Ok(n) => n,
            Err(err) => {
                println!("Error parsing number: {}", err);
                continue;
            }
        };

        sum += num;
    }

    println!("{}", sum);
}
